Signup: Refactor signup_domain_origin To Use Dependency Store

sharpirate opened the pull request:
Fixes 113-gh-Automattic/growth-foundations

### Summary
The logic for tracking the domain origin prop for the `calypso_signup_complete` event was implemented in this PR: [Signup: Add Domain Origin Prop to Tracks Event](https://github.com/Automattic/wp-calypso/pull/78085) which wasn't following best code practices because it used `localStorage` for tracking the data. This PR utilizes the dependency store of the signup framework to track that data across the steps.

### Testing Instructions
1. Checkout this branch
2. Run `yarn` and `yarn start`
3. Go to `http://calypso.localhost:3000/start`
4. Open [Tracks Vigilante](https://github.com/Automattic/tracks-chrome-extension) so you can monitor tracks events
5. Click on **Choose my domain later** in the sidebar. Continue through the flow until you see the `calypso_signup_complete` event fire in tracks vigilante. Confirm the even has the following property:
`signup_domain_origin: 'choose-later'`
6. Click on **Use a domain I own** in the sidebar. Continue through the flow until you see the `calypso_signup_complete` event fire in tracks vigilante. Confirm the even has the following property:
`signup_domain_origin: 'use-your-domain'`
7. Select the free .wordpress.com domain. Continue through the flow until you see the `calypso_signup_complete` event fire in tracks vigilante. Confirm the even has the following property:
`signup_domain_origin: 'free'`
8. Select the one of the custom suggested domains. Continue through the flow until you see the `calypso_signup_complete` event fire in tracks vigilante. Confirm the even has the following property:
`signup_domain_origin: 'custom'`
9. If for some reason the domain origin hasn't been set, the value of the `signup_domain_origin` should be `not-set`

### Important Note
We've already merged this functionality [here](https://github.com/Automattic/wp-calypso/pull/78962) but it was throwing type errors on production so we had to revert. In this PR I've included a safety type check for before calling `onConnect` and fixed the `domainName` prop not being passed down

sharpirate committed:
refactor domain origin logic to use dependency store instead of localStorage

sharpirate committed:
rename dependency

sharpirate committed:
handle use your own domain flow

sharpirate committed:
refactor signup domain origin type

sharpirate committed:
add comment for signup step check

sharpirate committed:
provide default value for domainOrigin outside of the signup flow

sharpirate committed:
safety type check for onConnect and forward domainName

sharpirate commented on client/components/domains/use-my-domain/index.jsx:
```diff
@@ -271,7 +271,13 @@ function UseMyDomain( props ) {
 						? showOwnershipVerificationFlow
-						: onConnect
+						: () =>
+								onConnect &&
```
safety type check before calling `onConnect`

sharpirate commented on client/components/domains/use-my-domain/index.jsx:
```diff
@@ -271,7 +271,13 @@ function UseMyDomain( props ) {
+						: () =>
+								onConnect &&
+								onConnect( {
+									domain: domainName,
```
pass down `domainName` which was missing in the previous PR

ddc22 commented on client/lib/analytics/signup.js:
```diff
@@ -25,6 +21,15 @@ export function recordSignupStart( flow, ref, optionalProps ) {
+	choose_later: 'choose-later',
+	free: 'free',
+	custom: 'custom',
+	not_set: 'not-set',
```
Please use camel case for variable names in JavaScript. If these are constants then they should be snake case, all caps. Generally snake case is used with php in the backend.

sharpirate replied:
Thanks, updated! 👍

alshakero commented on client/components/domains/use-my-domain/index.jsx:
```diff
@@ -271,7 +271,13 @@ function UseMyDomain( props ) {
 						? showOwnershipVerificationFlow
-						: onConnect
+						: () =>
+								onConnect &&
```
Can we make a function called `handleConnect` that will do all of this stuff? Having it inline is really prone to error.

sharpirate replied:
Following @jdc91's suggestion to avoid prop drilling because `handleAddMapping` is only being called when the user uses their own domain, I just set the `signupDomainOrigin` directly from within `handleAddMapping`. This in turn means we don't have to do the inline logic anymore

ddc22 commented on client/signup/steps/domains/index.jsx:
```diff
@@ -343,7 +338,11 @@ class DomainsStep extends Component {
+			Object.assign(
+				{ domainItem },
+				useThemeHeadstartItem,
+				signupDomainOrigin ? { signupDomainOrigin } : {}
```
I think we can prevent prop drilling by bringing the following logic directly here. `handleAddMapping` only happens when a domain is mapped. Same with `handleAddTransfer` happens only when a domain user has his own domain.

https://github.com/Automattic/wp-calypso/blob/4b67949dd83ab838875e6c5ffc8a4b47ae8bbf90/client/components/domains/use-my-domain/index.jsx#L279-L279

The isInSignup prop is actually hardcoded on this domain step so I don't think we should worry about it.

https://github.com/Automattic/wp-calypso/blob/4b67949dd83ab838875e6c5ffc8a4b47ae8bbf90/client/signup/steps/domains/index.jsx#L613

Do let me know if I am missing something here.

sharpirate replied:
That makes sense. I updated the PR to set the `signupDomainOrigin` directly within the `handleAddMapping ` so I can avoid prop drilling. This simplifies the PR a lot since I don't need the whole `onConnect` logic

ddc22 reviewed (changes requested):
Since the previous change lead to breaking change in production, we should revisit and do a root cause analysis to see what exactly the bug was to turn it into a learning opportunity :). Normally we would P2 and do a post mortem but since we caught this bug pretty early that is not required.

However, can you go into deeper detail as to what happened? What exactly was the cause of the breakage and how you have fixed that here (with a little bit more details). And possibly future steps we can take to prevent such an error from happening?

sharpirate committed:
address comments

Sharpirate commented:
>However, can you go into deeper detail as to what happened? What exactly was the cause of the breakage and how you have fixed that here (with a little bit more details). And possibly future steps we can take to prevent such an error from happening?

We were getting a **TypeError N is not a function** for the `onConnect` call. Here's the [Sentry log](https://a8c.sentry.io/issues/4307688773/?project=6313676). I don't quite understand how my previous changes could've caused it, my guess is it had to something with the `onConnect` function being out of scope due to the added anonymous inline function. The weird part is I couldn't reproduce the issue on local calypso.  Now with the new approach of setting the `signupDomainOrigin` directly from within `handleAddMapping` I've removed all the logic that modifies the `onConnect` callback so now it's virtually the same as in production. I've added a safety check `onConnect &&` before executing the callback although that's probably not needed but I think it's good to have because the function is coming from a prop and we're blindly executing it hoping it's defined.

ddc22 commented on client/components/domains/use-my-domain/index.jsx:
```diff
@@ -271,7 +270,7 @@ function UseMyDomain( props ) {
 					'auth_code' === domainAvailabilityData?.ownership_verification_type
 						? showOwnershipVerificationFlow
-						: onConnect
+						: onConnect && onConnect
```
Pending cleanup probably?

ddc22 replied:
This can lead to a breakage so make sure to clean this up. @Sharpirate

ddc22 commented on client/signup/steps/domains/index.jsx:
```diff
@@ -343,7 +338,9 @@ class DomainsStep extends Component {
 			),
-			Object.assign( { domainItem }, useThemeHeadstartItem )
+			Object.assign( { domainItem }, useThemeHeadstartItem, {
+				signupDomainOrigin: SIGNUP_DOMAIN_ORIGIN.USE_YOUR_DOMAIN,
```
Can you check with analytics whether we need to add this to domain transfers as well?

ddc22 commented on client/signup/steps/domains/index.jsx:
```diff
@@ -310,7 +304,8 @@ class DomainsStep extends Component {
 				this.isDependencyShouldHideFreePlanProvided() ? { shouldHideFreePlan } : {},
-				useThemeHeadstartItem
+				useThemeHeadstartItem,
+				signupDomainOrigin ? { signupDomainOrigin } : {}
```
Nit: can probably be simplified with.

```suggestion
				...( signupDomainOrigin && { signupDomainOrigin } ) 
```

ddc22 reviewed (approved):
I have explored the previously the domain mapping flow seems to be working. And for the rest the various prop seems to propagate through the steps. And the signup complete step successfully shows the relevant prop.

 I have left some minor code comments below.
 
Once resolved and shipped to the `staging environment`, my advice is to reach out to jp-imagines or anyone who could reproduce the previous bug and test out the domain connect feature and make sure there are no breaking changes. 

<img width="1626" alt="image" src="https://github.com/Automattic/wp-calypso/assets/3422709/ce88c5e6-88a3-471c-bb33-be98598a2156">

sharpirate committed:
clean up
